### ISSUE TITLE
An InstCombine patch is making the loop terminating condition more complicated

### ISSUE THREAD

**Issue by kazutakahirata** [llvm/llvm-project#54558]
I'm seeing a case where https://github.com/llvm/llvm-project/commit/61580d0949fd3465f53c71f5a8f304d4722a38fb is making the loop terminating condition more complicated.

Consider:
```
define i32 @foo(i32 %arg) {
bb_entry:
  br label %bb_outer

bb_outer:                                         ; preds = %bb_inner, %bb_entry
  %phi_outer = phi i32 [ %arg, %bb_entry ], [ %sub, %bb_inner ]
  %cond_outer = icmp ne i32 %phi_outer, 0
  br i1 %cond_outer, label %bb_preheader, label %bb_exit

bb_preheader:                                     ; preds = %bb_outer
  %ugt = icmp ugt i32 43, %phi_outer
  %sel = select i1 %ugt, i32 %phi_outer, i32 43
  %sub = sub i32 %phi_outer, %sel
  br label %bb_inner

bb_inner:                                         ; preds = %bb_inner, %bb_preheader
  %phi_inner = phi i32 [ %sel, %bb_preheader ], [ %iv, %bb_inner ]
  call void @bar()
  %iv = add i32 %phi_inner, 1
  %cond_inner = icmp ne i32 %phi_inner, 0
  br i1 %cond_inner, label %bb_inner, label %bb_outer

bb_exit:                                          ; preds = %bb_outer
  ret i32 0
}

declare void @bar()
```
Compile this like so:
```
$ clang -O2 -S reduced.ll
```
After InstCombinePass, I see a diff like so:
```
 *** IR Dump After InstCombinePass on foo ***
 define i32 @foo(i32 %arg) local_unnamed_addr {
 bb_entry:
   %cond_outer.not1 = icmp eq i32 %arg, 0
   br i1 %cond_outer.not1, label %bb_exit, label %bb_preheader
 
 bb_outer.loopexit:                                ; preds = %bb_inner
-  %cond_outer.not = icmp eq i32 %sub, 0
+  %cond_outer.not = icmp ult i32 %phi_outer2, 44
   br i1 %cond_outer.not, label %bb_exit, label %bb_preheader
 
 bb_preheader:                                     ; preds = %bb_entry, %bb_outer.loopexit
   %phi_outer2 = phi i32 [ %sub, %bb_outer.loopexit ], [ %arg, %bb_entry ]
   %0 = tail call i32 @llvm.umin.i32(i32 %phi_outer2, i32 43)
   %sub = sub i32 %phi_outer2, %0
   br label %bb_inner
 
 bb_inner:                                         ; preds = %bb_inner, %bb_preheader
   %phi_inner = phi i32 [ %0, %bb_preheader ], [ %iv, %bb_inner ]
   tail call void @bar()
   %iv = add i32 %phi_inner, 1
   %cond_inner.not = icmp eq i32 %phi_inner, 0
   br i1 %cond_inner.not, label %bb_outer.loopexit, label %bb_inner
 
 bb_exit:                                          ; preds = %bb_outer.loopexit, %bb_entry
   ret i32 0
 }
```
Notice that `icmp eq i32 ..., 0` is replaced with `icmp ult i32 ..., 44`.  This one-liner diff persists all the way through the remaining LLVM IR passes.

At the assembly level, I see a diff like so:
```
 foo:                                    # @foo
 # %bb.0:                                # %bb_entry
 	pushq	%rbp
 	pushq	%r14
 	pushq	%rbx
 	testl	%edi, %edi
 	je	.LBB0_5
 # %bb.1:
 	movl	%edi, %ebx
 	.p2align	4, 0x90
 .LBB0_3:                                # %bb_preheader
                                         # =>This Loop Header: Depth=1
                                         #     Child Loop BB0_4 Depth 2
 	cmpl	$43, %ebx
 	movl	$43, %r14d
 	cmovbl	%ebx, %r14d
 	leal	-1(%r14), %ebp
 	.p2align	4, 0x90
 .LBB0_4:                                # %bb_inner
                                         #   Parent Loop BB0_3 Depth=1
                                         # =>  This Inner Loop Header: Depth=2
 	callq	bar@PLT
 	incl	%ebp
 	jne	.LBB0_4
 # %bb.2:                                # %bb_outer.loopexit
                                         #   in Loop: Header=BB0_3 Depth=1
-	subl	%r14d, %ebx
-	jne	.LBB0_3
+	movl	%ebx, %eax
+	subl	%r14d, %eax
+	cmpl	$44, %ebx
+	movl	%eax, %ebx
+	jae	.LBB0_3
 .LBB0_5:                                # %bb_exit
 	xorl	%eax, %eax
 	popq	%rbx
 	popq	%r14
 	popq	%rbp
 	retq
```
Notice that we have three more instructions.

**Comment by kazutakahirata** [comment]
@nikic, do you think there is something quick you can do here?  If not, assign back to me.  Thanks!

**Comment by kazutakahirata** [comment]
The patch in question https://github.com/llvm/llvm-project/commit/61580d0949fd3465f53c71f5a8f304d4722a38fb says the motivating case is to recognize that the `select` below (taken from `llvm/test/Transforms/InstCombine/icmp-sub.ll`) can be simplified to `%b`:
```
define i32 @sub_eq_zero_select(i32 %a, i32 %b, i32* %p) {
; CHECK-LABEL: @sub_eq_zero_select(
; CHECK-NEXT:    [[SUB:%.*]] = sub i32 [[A:%.*]], [[B:%.*]]
; CHECK-NEXT:    store i32 [[SUB]], i32* [[P:%.*]], align 4
; CHECK-NEXT:    ret i32 [[B]]
;
  %sub = sub i32 %a, %b
  store i32 %sub, i32* %p
  %cmp = icmp eq i32 %sub, 0
  %sel = select i1 %cmp, i32 %a, i32 %b
  ret i32 %sel
}
```

Here are some options to fix this bug without breaking the motivating case above:
- Revert the patch, and teach the inst-combine to look into `icmp` and `sub` when it examines `select`.
- Keep the patch, and teach the backend to recognize `icmp ult i32 %phi_outer, 44` and transform it back to `icmp eq i32 %sub, 0` during the instruction selection or one of the DAG transformations.  This may be a bit complicated because `icmp ult ...` is a sibling node of `sub i32 %phi_outer, %min`.  That is, one is not a parent of the other.
- Keep the patch, and teach the inst-combine to refrain from transforming `icmp eq i32 %phi_outer, %min` into `icmp ult i32 %phi_outer, 44` when the conditional branch uses `icmp` and `sub i32 %phi_outer, %min` is available.  Again, this may be complicated because `icmp eq i32 %phi_outer, %min` has already become a sibling of `sub`.  That is, one is not a parent of the other.

**Comment by nikic** [comment]
> Revert the patch, and teach the inst-combine to look into `icmp` and `sub` when it examines `select`.

The problem with this is that it doesn't compose -- we can't reasonably teach all places that perform `icmp eq` folds that `icmp eq sub 0` is an equivalent pattern. We rely on InstCombine canonicalization to keep this kind of combinatorial explosion of patterns in check.

As far as I can see, the main problem here is not just the transform from `icmp eq i32 %sub, 0` to `icmp eq i32 %phi_outer, %min` -- at that point we'd still be able to easily reuse sub flags in the backend. However, this transform enables another transform to `icmp ult i32 %phi_outer, 44` which obscures the relationship between the icmp and the sub. This is nominally a desirable middle-end transform (because e.g. now range-based reasoning on outer_phi will work), but ends up being unfavorable for the backend, in a way it cannot recover.

I don't have any really great ideas on how to address this. Maybe @rotateright has some idea.

**Comment by rotateright** [comment]
If we can't rely on the backend to recover, then I would carve out an exception to the transform in IR (check if the other use of the sub is in a phi or something like that?). It's ugly, but it's the only practical way to get the code we want.

There is 1 more possibility for this example though: there's another one-use restriction in play that isn't strictly necessary. If you apply the patch below, does performance get better or worse in the motivating application?

```
diff --git a/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp b/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
index 5e4d842dd8e7..f2ba2697aad6 100644
--- a/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
+++ b/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
@@ -2152,7 +2152,7 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
         I, Builder.CreateIntrinsic(Intrinsic::usub_sat, {Ty}, {X, Op1}));
 
   // Op0 - umin(X, Op0) --> usub.sat(Op0, X)
-  if (match(Op1, m_OneUse(m_c_UMin(m_Value(X), m_Specific(Op0)))))
+  if (match(Op1, m_c_UMin(m_Value(X), m_Specific(Op0))))
     return replaceInstUsesWith(
         I, Builder.CreateIntrinsic(Intrinsic::usub_sat, {Ty}, {Op0, X}));
 
diff --git a/llvm/test/Transforms/InstCombine/sub-minmax.ll b/llvm/test/Transforms/InstCombine/sub-minmax.ll
index 2deca35cdc9f..ad8a42e5d03e 100644
--- a/llvm/test/Transforms/InstCombine/sub-minmax.ll
+++ b/llvm/test/Transforms/InstCombine/sub-minmax.ll
@@ -61,8 +61,8 @@ define i5 @sub_umin_uses(i5 %a, i5 %b, i5* %p) {
 ; CHECK-LABEL: @sub_umin_uses(
 ; CHECK-NEXT:    [[UMIN:%.*]] = call i5 @llvm.umin.i5(i5 [[A:%.*]], i5 [[B:%.*]])
 ; CHECK-NEXT:    store i5 [[UMIN]], i5* [[P:%.*]], align 1
-; CHECK-NEXT:    [[R:%.*]] = sub i5 [[A]], [[UMIN]]
-; CHECK-NEXT:    ret i5 [[R]]
+; CHECK-NEXT:    [[TMP1:%.*]] = call i5 @llvm.usub.sat.i5(i5 [[A]], i5 [[B]])
+; CHECK-NEXT:    ret i5 [[TMP1]]
 ;
   %umin = call i5 @llvm.umin.i5(i5 %a, i5 %b)
   store i5 %umin, i5* %p
@@ -253,10 +253,10 @@ define i32 @na_minus_max_na_bi_use2(i32 %A, i32 %Bi) {
 ; CHECK-LABEL: @na_minus_max_na_bi_use2(
 ; CHECK-NEXT:    [[NOT:%.*]] = xor i32 [[A:%.*]], -1
 ; CHECK-NEXT:    [[TMP1:%.*]] = call i32 @llvm.umin.i32(i32 [[NOT]], i32 31)
-; CHECK-NEXT:    [[X:%.*]] = sub i32 [[NOT]], [[TMP1]]
+; CHECK-NEXT:    [[TMP2:%.*]] = call i32 @llvm.usub.sat.i32(i32 [[NOT]], i32 31)
 ; CHECK-NEXT:    call void @use32(i32 [[TMP1]])
 ; CHECK-NEXT:    call void @use32(i32 [[NOT]])
-; CHECK-NEXT:    ret i32 [[X]]
+; CHECK-NEXT:    ret i32 [[TMP2]]
 ;
   %not = xor i32 %A, -1
   %l0 = icmp ult i32 %not, 31
@@ -472,8 +472,8 @@ define i8 @umin_not_sub_intrinsic_uses(i8 %x, i8 %y) {
 ; CHECK-NEXT:    call void @use8(i8 [[NY]])
 ; CHECK-NEXT:    [[M:%.*]] = call i8 @llvm.umin.i8(i8 [[NX]], i8 [[NY]])
 ; CHECK-NEXT:    call void @use8(i8 [[M]])
-; CHECK-NEXT:    [[SUBX:%.*]] = sub i8 [[NX]], [[M]]
-; CHECK-NEXT:    ret i8 [[SUBX]]
+; CHECK-NEXT:    [[TMP1:%.*]] = call i8 @llvm.usub.sat.i8(i8 [[NX]], i8 [[NY]])
+; CHECK-NEXT:    ret i8 [[TMP1]]
 ;
   %nx = xor i8 %x, -1
   call void @use8(i8 %nx)
@@ -599,8 +599,8 @@ define i8 @umin_sub_op1_use(i8 %x, i8 %y) {
 ; CHECK-LABEL: @umin_sub_op1_use(
 ; CHECK-NEXT:    [[U:%.*]] = call i8 @llvm.umin.i8(i8 [[Y:%.*]], i8 [[X:%.*]])
 ; CHECK-NEXT:    call void @use8(i8 [[U]])
-; CHECK-NEXT:    [[R:%.*]] = sub i8 [[Y]], [[U]]
-; CHECK-NEXT:    ret i8 [[R]]
+; CHECK-NEXT:    [[TMP1:%.*]] = call i8 @llvm.usub.sat.i8(i8 [[Y]], i8 [[X]])
+; CHECK-NEXT:    ret i8 [[TMP1]]
 ;
   %u = call i8 @llvm.umin.i8(i8 %y, i8 %x)
   call void @use8(i8 %u)

```

**Comment by kazutakahirata** [comment]
@rotateright Thank you for your patch.  Unfortunately, it doesn't help.  I get:

```
bb_loop:                                          ; preds = %bb_loop, %bb_entry
  %phi_outer = phi i32 [ %0, %bb_loop ], [ %arg, %bb_entry ]
  %min = tail call i32 @llvm.umin.i32(i32 %phi_outer, i32 43)
  tail call void @bar(i32 %min)
  %0 = tail call i32 @llvm.usub.sat.i32(i32 %phi_outer, i32 43)
  %cond_outer = icmp ult i32 %phi_outer, 44
  br i1 %cond_outer, label %bb_exit, label %bb_loop
```
which ends up with:
```
.LBB0_1:                                # %bb_loop
                                        # =>This Inner Loop Header: Depth=1
        cmpl    $43, %ebx
        movl    $43, %edi
        cmovbl  %ebx, %edi
        callq   bar@PLT
        subl    $43, %ebx
        cmovbl  %ebp, %ebx
        ja      .LBB0_1
```

**Comment by kazutakahirata** [comment]
@rotateright Thanks!  Your patch works.  May I ask why you are checking the PHI node (as opposed to umin or other hard-to-remove uses like function arguments or memory stores)?  Is that because we probably cannot optimize away the SUB if its result is used in a PHI node?  And, if the SUB stays, we compute flags for free?

**Comment by rotateright** [comment]
We're inventing a heuristic based on your example code. :)
The problem pattern (and also likely for the other folds just below the one that we are changing - see the TODO comment) seems to be when the subtract is used in a loop condition, so I'm just trying to create an exception for that pattern. 

It's entirely possible that the proposed one-liner change is too big a hammer, too small, or both. But I think it's ok to start with that and refine. We at least know it won't regress the motivating example from 61580d0949fd3465f53c71f5a8.

**Comment by kazutakahirata** [comment]
@rotateright I see.  Would you mind check in your patch?  Thanks in advance!

**Comment by rotateright** [comment]
Added tests/comments and posted for review:
https://reviews.llvm.org/D122909

**Comment by nikic** [comment]
Closing this as the aforementioned patch has landed.

**Comment by rotateright** [comment]
I think we can close this report now - fixed by 5f8c2b884d4288  .
Please reopen or file another bug if we there are still known problems.
